src/equations/numerical_fluxes.jl

```diff
@@ -53,19 +53,44 @@ struct FluxRotated{NumericalFlux}
   numerical_flux::NumericalFlux
 end
 
-@inline function (flux_rotated::FluxRotated)(u_ll, u_rr, normal_vector::AbstractVector, equations)
+@inline function (flux_rotated::FluxRotated)(u_ll, u_rr, normal, equations::AbstractEquations{2})
   @unpack numerical_flux = flux_rotated
 
-  norm_ = norm(normal_vector)
-  # normalize the vector without using `normalize` since we need to multiply by the `norm_` later
-  normal = normal_vector / norm_
+  norm_ = norm(normal)
+  # Normalize the vector without using `normalize` since we need to multiply by the `norm_` later
+  normal_normalized = normal / norm_
 
-  u_ll_rotated = rotate_to_x(u_ll, normal, equations)
-  u_rr_rotated = rotate_to_x(u_rr, normal, equations)
+  u_ll_rotated = rotate_to_x(u_ll, normal_normalized, equations)
+  u_rr_rotated = rotate_to_x(u_rr, normal_normalized, equations)
 
   f = numerical_flux(u_ll_rotated, u_rr_rotated, 1, equations)
 
-  return rotate_from_x(f, normal, equations) * norm_
+  return rotate_from_x(f, normal_normalized, equations) * norm_
+end
+
+
+@inline function (flux_rotated::FluxRotated)(u_ll, u_rr, normal, equations::AbstractEquations{3})
+  @unpack numerical_flux = flux_rotated
+
+  norm_ = norm(normal)
+  # Normalize the vector without using `normalize` since we need to multiply by the `norm_` later
+  normal_normalized = normal / norm_
+
+  # Some vector that can't be identical to normal (unless normal == 0)
+  tangent1 = SVector(normal[2], normal[3], -normal[1])
+  # Orthogonal projection
+  tangent1 -= dot(normal_normalized, tangent1) * normal_normalized
+  tangent1 = normalize(tangent1)
+
+  # Third orthogonal vector
+  tangent2 = normalize(cross(normal, tangent1))
```

Comment:
Okay, we implemented it with storing the normal and tangent vectors in the `element_container`.
For `elixir_euler_soruce_terms` in 3D we now get these results with storing vectors:
```julia

 ────────────────────────────────────────────────────────────────────────────
           Trixi.jl                   Time                   Allocations
                              ──────────────────────   ───────────────────────
       Tot / % measured:           952ms / 97.4%           4.87MiB / 86.8%

 Section              ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 rhs!                  1.59k    867ms  93.5%   547μs    897KiB  20.7%     579B
   interface flux      1.59k    315ms  33.9%   198μs     0.00B  0.00%    0.00B
   volume integral     1.59k    284ms  30.6%   179μs     0.00B  0.00%    0.00B
   source terms        1.59k    127ms  13.7%  80.1μs    446KiB  10.3%     288B
   surface integral    1.59k    107ms  11.5%  67.3μs     0.00B  0.00%    0.00B
   Jacobian            1.59k   29.3ms  3.16%  18.5μs     0.00B  0.00%    0.00B
   reset ∂u/∂t         1.59k   3.05ms  0.33%  1.93μs     0.00B  0.00%    0.00B
   ~rhs!~              1.59k   2.14ms  0.23%  1.35μs    451KiB  10.4%     291B
 calculate dt            318   25.7ms  2.77%  80.7μs     0.00B  0.00%    0.00B
 analyze solution          5   18.6ms  2.01%  3.73ms    286KiB  6.61%  57.3KiB
 I/O                      11   16.1ms  1.74%  1.46ms   3.07MiB  72.7%   286KiB
 ────────────────────────────────────────────────────────────────────────────
```
and with the old version:
```julia
 ────────────────────────────────────────────────────────────────────────────
           Trixi.jl                   Time                   Allocations
                              ──────────────────────   ───────────────────────
       Tot / % measured:           1.05s / 97.5%           4.87MiB / 86.8%

 Section              ncalls     time   %tot     avg     alloc   %tot      avg
 ────────────────────────────────────────────────────────────────────────────
 rhs!                  1.59k    958ms  93.7%   604μs    897KiB  20.7%     579B
   interface flux      1.59k    401ms  39.2%   253μs     0.00B  0.00%    0.00B
   volume integral     1.59k    287ms  28.0%   181μs     0.00B  0.00%    0.00B
   source terms        1.59k    129ms  12.6%  81.1μs    446KiB  10.3%     288B
   surface integral    1.59k    108ms  10.6%  68.4μs     0.00B  0.00%    0.00B
   Jacobian            1.59k   27.3ms  2.67%  17.2μs     0.00B  0.00%    0.00B
   reset ∂u/∂t         1.59k   2.99ms  0.29%  1.88μs     0.00B  0.00%    0.00B
   ~rhs!~              1.59k   2.45ms  0.24%  1.54μs    451KiB  10.4%     291B
 calculate dt            318   26.0ms  2.55%  81.9μs     0.00B  0.00%    0.00B
 analyze solution          5   19.5ms  1.91%  3.91ms    286KiB  6.61%  57.3KiB
 I/O                      11   19.4ms  1.89%  1.76ms   3.07MiB  72.7%   286KiB
 ───────────────────────────────────────────────────────────────────────────
```
When comparing `calc_interface_flux!`, we get the following results:
with storing vectors:
```julia
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     192.400 μs (0.00% GC)
  median time:      194.000 μs (0.00% GC)
  mean time:        196.055 μs (0.00% GC)
  maximum time:     623.400 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
and without storing:
```julia
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     247.800 μs (0.00% GC)
  median time:      248.400 μs (0.00% GC)
  mean time:        250.218 μs (0.00% GC)
  maximum time:     370.100 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```
We used `FluxLaxFriedrichs`. For the new implementation we need to pass the tangent vectors for the flux calculation, so we have to dispatch by flux type again.

Reply:
:+1:

Reply:
So, should we leave it as is or should we store the vectors for a bit better performance?

Reply:
Ah, I think I misunderstood. I thought you already had changed the implementation. I think it is fine to keep the current implementation for now, but maybe leave a comment in there (in `calc_interface_flux!`) that explains that caching the tangent vectors improves the performance of `calc_interface_flux!` by about 20%. This way, we know that this is something we can optimize later on should performance becomes a bottleneck.

You could consider creating a function `tangent1, tangent2 = calc_tangent_vectors(normal, normal_normalized)` that calculates the tangent vectors and that can be used elsewhere (and benchmarked individually), but it is not absolutely necessary.

You could even make it `calc_tangent_vectors(normal_normalized)` if you start with
```julia
tangent1 = SVector(normal_normalized[2], normal_normalized[3], -normal_normalized[1])
```
Independent of the remaining comment: If you compute `tangent2` as
```julia
tangent2 = cross(normal_normalized, tangent1)
```
you don't have to normalize it, since the cross product of two unit vectors is already one. That should shave off some more time.